cypher-shell/src/main/java/org/neo4j/shell/Main.java

```diff
@@ -104,7 +104,12 @@ void connectInteractively(@Nonnull CypherShell shell, @Nonnull ConnectionConfig
         try {
             shell.connect(connectionConfig);
         } catch (ClientException e) {
-            if (e.getMessage() == null || !e.getMessage().contains("Missing username")) {
+            // Only prompt for username/password if they weren't used
+            if (!connectionConfig.username().isEmpty() && !connectionConfig.password().isEmpty()) {
+                throw e;
+            }
+            // Errors except authentication related should be shown to user
+            if (e.code() == null || !e.code().contains("Unauthorized")) {
```

Comment:
The codes are part of Neo4j's client API. You can reasonably depend on them not changing, so you could do a full equality check here.